iml-systemd/src/systemd_error.rs

```diff
@@ -0,0 +1,95 @@
+use iml_cmd::CmdError;
+use std::{fmt, process::Output};
+
+#[derive(Debug)]
+pub struct RequiredError(pub String);
+
+impl fmt::Display for RequiredError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self.0)
+    }
+}
+
+impl std::error::Error for RequiredError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        None
+    }
+}
+
+#[derive(Debug)]
+pub enum SystemdError {
+    Io(std::io::Error),
+    Utf8Error(std::str::Utf8Error),
+    RequiredError(RequiredError),
+    CmdError(CmdError),
+    UnexpectedStatusError,
+}
+
+impl std::fmt::Display for SystemdError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match *self {
+            SystemdError::Io(ref err) => write!(f, "{}", err),
+            SystemdError::Utf8Error(ref err) => write!(f, "{}", err),
+            SystemdError::RequiredError(ref err) => write!(f, "{}", err),
+            SystemdError::CmdError(ref err) => write!(f, "{}", err),
+            SystemdError::UnexpectedStatusError => write!(f, "Unexpected status code"),
+        }
+    }
+}
+
+impl std::error::Error for SystemdError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match *self {
+            SystemdError::Io(ref err) => Some(err),
+            SystemdError::Utf8Error(ref err) => Some(err),
+            SystemdError::RequiredError(ref err) => Some(err),
+            SystemdError::CmdError(ref err) => Some(err),
+            SystemdError::UnexpectedStatusError => None,
+        }
+    }
+}
+
+impl From<SystemdError> for std::io::Error {
```

Comment:
This conversion seems odd, I don't think we need this.

Reply:
Let me see what happens if I remove it.

Reply:
Seems like it's ok

Reply:
The issue I'm running into at the moment is here:
```
rror[E0277]: `?` couldn't convert the error to `std::io::Error`
  --> src/docker.rs:42:62
   |
42 |     iml_systemd::stop_unit("iml-docker.service".into()).await?;
   |                                                              ^ the trait `std::convert::From<iml_systemd::systemd_error::SystemdError>` is not implemented for `std::io::Error`
```

Reply:
yeah, `remove_iml_stack` (the code where that's being called) should be like `pub async fn remove_iml_stack() -> Result<(), iml_systemd::SystemdError> {`

There is no need for it to return `io::Error`.

Reply:
I was thinking I could just map the error, something like this:
```
iml_systemd::stop_unit("iml-docker.service".into()).await
      .map_err(|x| io::Error::new(io::ErrorKind::Other, x.to_string()))?;
```
I thought we would still need the error type to be io::Error since we have a couple of commands that follow. For example:
```
x.arg("stack").arg("rm").arg("iml").checked_status().await?;
```
Those will use io::Error won't they?

Reply:
> Those will use io::Error won't they? 

Yes, but `SystemdError` wraps io::Error, so `?` will cast appropriately.

Reply:
nvm,  looks like it compiles just fine.

Reply:
ah that makes sense